src/pessimisticPatternMatcherUtils.cc

```diff
@@ -0,0 +1,115 @@
+        unsigned int ref_id = ref_id_array[*idx];
+        double ref_delta[3];
+        double proj_ref_delta[3];
+        ref_delta[0] = reference_array[ref_id][0] - ref_ctr[0];
+        ref_delta[1] = reference_array[ref_id][1] - ref_ctr[1];
+        ref_delta[2] = reference_array[ref_id][2] - ref_ctr[2];
+
+        double ref_dot = ref_delta[0] * ref_ctr[0] + ref_delta[1] * ref_ctr[1] + ref_delta[2] * ref_ctr[2];
+        proj_ref_delta[0] = ref_delta[0] - ref_dot * ref_ctr[0];
+        proj_ref_delta[1] = ref_delta[1] - ref_dot * ref_ctr[1];
+        proj_ref_delta[2] = ref_delta[2] - ref_dot * ref_ctr[2];
+        /// Compute the cos between our "center" reference vector and the
+        /// current reference candidate.
+        double proj_delta_dist_sq =
+                (proj_ref_delta[0] * proj_ref_delta[0] + proj_ref_delta[1] * proj_ref_delta[1] +
+                 proj_ref_delta[2] * proj_ref_delta[2]);
+        double geom_dist_ref = sqrt(proj_ref_ctr_dist_sq * proj_delta_dist_sq);
+        double cos_theta_ref =
+                (proj_ref_delta[0] * proj_ref_ctr_delta[0] + proj_ref_delta[1] * proj_ref_ctr_delta[1] +
+                 proj_ref_delta[2] * proj_ref_ctr_delta[2]) /
+                geom_dist_ref;
+
+        /// Make sure we can safely make the comparison in case
+        /// our "center" and candidate vectors are mostly aligned.
+        double cos_sq_comparison;
+        if (cos_theta_ref * cos_theta_ref < (1 - src_sin_tol * src_sin_tol)) {
+            cos_sq_comparison = (cos_theta_src - cos_theta_ref) * (cos_theta_src - cos_theta_ref) /
+                                (1 - cos_theta_ref * cos_theta_ref);
+        } else {
+            cos_sq_comparison = (cos_theta_src - cos_theta_ref) * (cos_theta_src - cos_theta_ref) /
+                                (src_sin_tol * src_sin_tol);
+        }
+        /// Test the difference of the cosine of the reference angle against
+        /// the source angle. Assumes that the delta between the two is
+        /// small.
+        if (cos_sq_comparison > src_sin_tol * src_sin_tol) {
+            continue;
+        }
+        /// The cosine tests the magnitude of the angle but not
+        /// its direction. To do that we need to know the sine as well.
+        /// This cross product calculation does that.
+        double cross_ref[3];
+        cross_ref[0] =
+                (proj_ref_delta[1] * proj_ref_ctr_delta[2] - proj_ref_delta[2] * proj_ref_ctr_delta[1]) /
+                geom_dist_ref;
+        cross_ref[1] =
+                (proj_ref_delta[2] * proj_ref_ctr_delta[0] - proj_ref_delta[0] * proj_ref_ctr_delta[2]) /
+                geom_dist_ref;
+        cross_ref[2] =
+                (proj_ref_delta[0] * proj_ref_ctr_delta[1] - proj_ref_delta[1] * proj_ref_ctr_delta[0]) /
+                geom_dist_ref;
+        double sin_theta_ref =
+                cross_ref[0] * ref_ctr[0] + cross_ref[1] * ref_ctr[1] + cross_ref[2] * ref_ctr[2];
```

Comment:
And again, with Eign, something like:

```
double sin_theta_ref = cross_ref.asEigenMatrix().dot(ref_ctr.asEigenMatrix());
```

Reply:
I'll give these a shot on one or two of them and see if I can get it to compile/run. If I can't though, I'm not too uncomfortable with just leaving it for now.

Reply:
Done.